### DISCUSSION TITLE
Just joined the obs-ndi team to help out; here are my plans

### DISCUSSION THREAD

**Discussion by paulpv** [DistroAV/DistroAV#830] [General]
@Palakis and @tt2468 recently added me to the `obs-ndi` team to help out.

I do like where the `rewrite` branch was going and how clean it was, but it was a fairly major change, even if "necessary", and I think most people (myself included) just want to start off with the old plugin + bug fixes on OBS 28+ and as few other changes as possible....and then optionally segue toward the rewrite version when we are ready.

My current plans:
1. Get this repo building DDRBoxman's https://github.com/Palakis/obs-ndi/pull/776 and officially release it here:
    1. `rewrite` is already currently the default branch
    1. DONE: ~~Branch a `pre-obs28` from `master` [for OBS27 legacy users]~~
    1. DONE: ~~Merge DDRBoxman's https://github.com/Palakis/obs-ndi/pull/776 to `master`~~
    1. DONE: ~~Make `master` the default branch again~~
    1. Release 4.10.2
1. DONE: ~~Simple merge of one liner https://github.com/Palakis/obs-ndi/pull/755 to `pre-obs28`~~
1. Automated builds and releases of separate binaries of `pre-obs28`, `master`, and `rewrite` branches
1. Over time converge `master` and `rewrite`
1. **Triage some of the bugs/Issues and resolve most or maybe even all of them and start from scratch.**
1. Whatever else issues uncovered while triaging Issues or discussing between us or on the forum(s).
1. Try to get an ARM64 build working for both Windows and MacOS.
  I have a M1 Macbook and Windows Devkit 2023 to help me test this.
  I am not sure yet if NDI Tools for MacOS even ships a native ARM64 build or what their plans are here.
  Maybe I can reach out to them and get a conversation going or an early preview.

What do y'all think, including @tt2468 ?

**Comment by Trouffman** [comment]
I like the ideas!
I did a couple of follow-up int he discussion / bug reports. a LOT of triage is required.

5. > Happy to jump onboard and be more involved on that part (including setting a process for reporting bug / challenges) and also a major "How to debug/test before reporting" becasue this lead to a lot of bloat.
We need a couple tags to sort through Win / Linux / MacOS based issues. Also a tag for "Legacy" and Post-V28-related Branch, and later have the "V5" approach maybe.

6. > Some of the bugs reported are very old or not applicable anymore.

7. > NDI Contact, I can help with this. 
There is no ARM64 (Windows-oriented) NDI Tools, AFAIK.

**Comment by paulpv** [comment]
An update on my progress.

Done:
1. Branched [pre-obs28](https://github.com/Palakis/obs-ndi/tree/pre-obs28) from [master](https://github.com/Palakis/obs-ndi/tree/master) to keep a pre-OBS28 option alive for people that want to continue to not update to OBS28 or above
2. Merged my [the one line audio fix](https://github.com/Palakis/obs-ndi/pull/755) in to `pre-obs28`
3. Branched [develop](https://github.com/Palakis/obs-ndi/tree/develop) from [master](https://github.com/Palakis/obs-ndi/tree/master) to have a working branch for main plugin improvements before merging back to `master`
4. Merged [DDRBoxman's obsplugin-template PR](https://github.com/Palakis/obs-ndi/pull/776) to `develop`
5. Created https://github.com/Palakis/obs-ndi/pull/834 to merge DDRBoxman's audio patch to `develop` (DDRBoxman's #776 PR did not have the audio fix)
6. Merged https://github.com/Palakis/obs-ndi/pull/834 to `develop`
7. Merged https://github.com/Palakis/obs-ndi/pull/835 to `develop` that enabled NDI5 on Linux (NDI5 already worked for MacOS and Windows).
8. Testing my [paulpv/obs-ndi/tree/develop](https://github.com/paulpv/obs-ndi/tree/develop) fork for MacOS, Ubuntu, and Windows.
    I am seeing that Windows and Linux work w/ NDI5, but for MacOS I am seeing the following problems:
      1. MacOS video output is very slow framerate (viewing on Windows NDI Tools NDI Studio Monitor)
      2. MacOS audio output is silent (viewing on Windows NDI Tools NDI Studio Monitor)
    **I have disabled MacOS builds until I resolve this.**  
    I may proceed Windows and Linux releases out the door to at least get most people unblocked.
    Until then you can get Windows and Linux builds from my branch's build actions, such as:  
    https://github.com/paulpv/obs-ndi/actions/runs/4149510017

Lots of things I still need to do:
1. Get Linux `.deb` installer working (I can get things working by copying files manually, but `apt install`ing the `.deb` file puts the files in a folder that OBS never sees :/)
2. Release new `libndi` install script for Linux builds.
3. Properly sign the Windows build so that Windows does not complain that it is untrusted or matches a `Trojan:Win32/Wacatac.H!ml` virus signature (I promise you that it is not a virus; there are a few other similar false reports for other apps out there).
4. Get `pre-obs28`, `master`, and `rewrite` branch build actions all working cooperatively.
5. Debug and fix why MacOS output has slow framerate and no audio.
6. Finish getting Raspberry Pi Bullseye x64 build working. :)

**Comment by paulpv** [comment]
Still making progress.

I am close to having a new obs-ndi ready for review and release.

Anyone have any opinions on:
1. Ideas why MacOS obs-ndi has no audio and < 1 FPS (when viewed with Windows NDI Studio Monitor)?
2. If it is worth the effort to try to get OBS28/29+obs-ndi to work on Raspberry Pi 4 [B 8GB]? Seems like that would be cool to have, but it also seems like OBS28+ does not directly support PiOS, so I may be wasting my time. Do you know if getting this to work would require full 64bit PiOS and OBS28+ and plugin (which I am assuming), or can there be any 32bit binary in this mix? I can get a full 64bit aarch64  linux OBS29+obsndi+libndi5 to compile and launch, but it immediately closes with "Aborted":
```
pv@raspberrypi:~/src/GitHub/obsproject/obs-studio $ MESA_GL_VERSION_OVERRIDE=3.3 obs -v
…
info: [obs-ndi] hello ! (version 4.10.1)
info: [obs-ndi] Trying ''
info: [obs-ndi] Trying '/usr/lib'
info: [obs-ndi] Trying '/usr/local/lib'
info: [obs-ndi] Trying '/usr/local/lib/aarch64-linux-gnu'
info: [obs-ndi] Found NDI library at '/usr/local/lib/aarch64-linux-gnu/libndi.so.5'
info: [obs-ndi] NDI runtime loaded successfully
info: [obs-ndi] NDI library initialized successfully (NDI SDK LINUX 09:57:33 Feb  2 2023 5.5.3)
QWidget: Must construct a QApplication before a QWidget
Aborted
pv@raspberrypi:~/src/GitHub/obsproject/obs-studio $ 
```
My compiled OBS29 runs fine if I disable obs-ndi.

**Comment by Trouffman** [comment]
Some opinions (as requested ) 
1. is the NDI Analysis tool any help to get some informations? - Is this the same when using Studio Monitor with he NDI Tools v4.x ?
2. If the pre-OBS28 /32bit of the combo used to work on Rpi (never tested) > why not relaease it (with a mention that support is limited / non-exsitant.).

Error : "QWidget: Must construct a QApplication before a QWidget"
That aim to Qt Version problem usually.
> Seems you are using the OBS-NDI version 4.10.1 > Did you compile it with Qt5(not even sure it's possible...) or Qt6 ? (that error would pop-up if you compliled with Qt5 IIRC)
> StremFX had a simialr problem in the past : https://github.com/Xaymar/obs-StreamFX/pull/428/commits/c4a98ed0386b070ff8fc8d236f052cb5155400dc Seems they fixed it by loading some UI stauff once OBS was fully loaded. Which would make sense for the error given. but not why this is only on your Rpi build...

(I would LOVE to have all this on a Pi :D)

**Reply by paulpv** [reply]
I think the `QWidget: Must construct a QApplication before a QWidget` is a red herring that always happens after there is any crash.

I will look in to the NDI Analysis tool.
I will compare/contrast to OBS 27 + NDI 4.

I do have plans to continue to release the legacy OBS 27 version, but could add to it the option to load either NDI 5 or fallback to NDI 4.

**Comment by jbcurler2010** [comment]
So thankful for the work that has been done and looking forward to a new release. Question: I posted this in a feature request some time ago. Is it possible to build in the option for NDI Output (main and or Dedicated filter) to use NDI HX?

**Reply by paulpv** [reply]
I think NDI|HX requires a different licensing model.
The source code for NDI|HX .h files says it is not to be distributed, so I am not sure how this project's open source [GNU 2.0 licensed](https://github.com/Palakis/obs-ndi/blob/master/LICENSE) code would be affected by that.
I am interested in looking in to that myself once I ship a basic version.
I will bug an NDI contact when I am ready for that.
